### ISSUE TITLE
segfault

### ISSUE THREAD

**Issue by schmurfy** [uptrace/uptrace-go#63]
I am not really sure how it happened but we had to rollback a service after upgrading uptrace-go, here is what was logged:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
```

and right after that:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xcb4b0a]

goroutine 1197 [running]:
go.opentelemetry.io/otel/sdk/trace.(*span).End(0xc000144a80, 0x0, 0x0, 0x0)
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v0.15.0/trace/span.go:158 +0x199
panic(0x145a380, 0x1faaa00)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/gogo/protobuf/types.(*Empty).MarshalToSizedBuffer(...)
	/go/pkg/mod/github.com/gogo/protobuf@v1.3.1/types/empty.pb.go:194
github.com/gogo/protobuf/types.(*Empty).Marshal(0x0, 0x15bb820, 0x0, 0x7f05cbc105f0, 0x0, 0x1)
	/go/pkg/mod/github.com/gogo/protobuf@v1.3.1/types/empty.pb.go:177 +0x6a
google.golang.org/protobuf/internal/impl.legacyMarshal(0x17c4880, 0xc000209ee0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000893200, 0x8595e2, ...)
	/go/pkg/mod/google.golang.org/protobuf@v1.25.0/internal/impl/legacy_message.go:391 +0xb0
google.golang.org/protobuf/proto.MarshalOptions.size(0xc000000000, 0x17c4880, 0xc000209ee0, 0x0)
	/go/pkg/mod/google.golang.org/protobuf@v1.25.0/proto/size.go:43 +0xcf
google.golang.org/protobuf/proto.MarshalOptions.Size(0x1000000, 0x177f040, 0xc000209ee0, 0xc000209ee0)
	/go/pkg/mod/google.golang.org/protobuf@v1.25.0/proto/size.go:26 +0x55
google.golang.org/protobuf/proto.Size(...)
	/go/pkg/mod/google.golang.org/protobuf@v1.25.0/proto/size.go:16
github.com/golang/protobuf/proto.Size(0x7f05cbc10580, 0x0, 0x0)
	/go/pkg/mod/github.com/golang/protobuf@v1.4.3/proto/wire.go:18 +0x69
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.messageType.Event(0x16174be, 0xc, 0x8, 0x0, 0x160d63a, 0x4, 0x0, 0x0, 0x17b1640, 0xc000460600, ...)
	/go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.15.1/interceptor.go:52 +0x20e
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1(0x17b1640, 0xc000460600, 0x15b70a0, 0xc0005bb6d0, 0xc00027cf00, 0xc00027d020, 0x0, 0x0, 0x0, 0x0)
	/go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.15.1/interceptor.go:341 +0xcc5
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x17b1640, 0xc0004601b0, 0x15b70a0, 0xc0005bb6d0, 0xc00098d900, 0x0, 0xc0006f2b30, 0x40e358)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.2.0/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x17b1640, 0xc0004601b0, 0x15b70a0, 0xc0005bb6d0, 0xc00027cf00, 0xc00027cf20, 0xc0006f2ba0, 0x54a106, 0x15343a0, 0xc0004601b0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.2.0/chain.go:34 +0xd7
github.com/birotaio/backend-services/generated_pb/bike._BikeService_SetAutoDiags_Handler(0x1606960, 0xc000124760, 0x17b1640, 0xc0004601b0, 0xc000509e00, 0xc000479c80, 0x17b1640, 0xc0004601b0, 0xc000589c30, 0xe)
	/tmp/build/80754af9/backend-services/bike/generated_pb/bike/bike.pb.go:16347 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000183c00, 0x17bfde0, 0xc0001a4180, 0xc00098d900, 0xc000479dd0, 0x1fc0078, 0x0, 0x0, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:1210 +0x522
google.golang.org/grpc.(*Server).handleStream(0xc000183c00, 0x17bfde0, 0xc0001a4180, 0xc00098d900, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:1533 +0xd05
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0002747a0, 0xc000183c00, 0x17bfde0, 0xc0001a4180, 0xc00098d900)
	/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:871 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:869 +0x1fd
```

The project is running uptrace-go version 0.6.1 and otel 0.15, the crashes occured on the 15/01 between 15:00 and 21:00 CET, we rollbacked the service and this morning I deployed it again with the same version and everything went fine for the whole day.

I am not sure whether uptrace or otel caused the crash but I don't know how I could trigger that crash from our code, especially since the crash occurs when the deferred `End` call happens.
No matter what happens the instrumentation should not be able to cause a segfault.

**Comment by vmihailenco** [comment]
Hi,

This looks like a problem with protobuf, not OpenTelemetry or uptrace-go . At go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.15.1/interceptor.go:52 I see:

```go
    semconv.RPCMessageUncompressedSizeKey.Int(proto.Size(p))
```

which just asks the size of the protobuf message and looks correct and harmless.

The real problem happens at https://github.com/protocolbuffers/protobuf-go/blob/v1.25.0/internal/impl/legacy_message.go#L385-L398. For some reason `protoUnwrap` returns typed nil interface (`(*types.Empty)(nil)`) and calling `Marshal` on it panics (`types.(*Empty).MarshalToSizedBuffer`).

Do you have something unusual in bike.pb.go that involves `types.Empty`?

---

If you don't mind I would report this issue at https://github.com/open-telemetry/opentelemetry-go-contrib with a workaround. Then you could add a `replace` directive to your go.mod file until Otel releases a fix (or says that it is a problem with protobuf :). But I recommend you to start updating to Otel v0.16 and uptrace-go v0.7.1, because otherwise you will need way more `replace` directives :)

**Comment by vmihailenco** [comment]
And the workaround I have in mind is https://github.com/uptrace/opentelemetry-go-contrib/commit/f82c0f8a19f7797228c8b06cc3f318eda628397b

**Comment by schmurfy** [comment]
I will try latest uptrace and otel and see if it persists first (as soon as pg and redis lib have updated to otel 0.16), I don't think we have something unusual in the generated code but we use gogopro not protofuf which may cause the issue, no idea.

the error happens only one one of our services but all are using the same dependencies.

**Comment by vmihailenco** [comment]
>as soon as pg and redis lib have updated to otel v0.16

I will bump dependencies in those libs, but AFAIK it is not needed. go mod will use the latest version anyway. I am actually surprised go mod still uses v0.15 in your project, because I had to use `replace` directive when I needed to test code on old Otel version...

>see if it persists first

Sure. To be clear I don't see how otel v0.16 can fix this issue, but it is worth trying.
